### PR TITLE
Refactor Metrics Key&Collector

### PR DESCRIPTION
**Pull request by wxbty**
## What is the purpose of the change
1. Metricskey optimization, directly dependent on Metricskey enum, remove ApplicationType/ServiceType type
2. Listener/Collector abstracts the corresponding App/Service type interface and abstract class, reduces repeated code as much as possible, and has a clearer object model
link #11941

### CODE REVIEW

**Review by AlbumenJ — APPROVED** [review]
LGTM
@songxiaosheng PTAL

File: dubbo-metrics/dubbo-metrics-api/src/main/java/org/apache/dubbo/metrics/listener/MetricsServiceListener.java [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+    }
+
+    public static AbstractMetricsListener onErrorEventBuild(MetricsKey metricsKey, MetricsPlaceType placeType, ServiceMetricsCollector<TimeCounterEvent> collector) {
+        return AbstractMetricsListener.onError(metricsKey,
+            event -> incrAndAddRt(metricsKey, placeType, collector, event)
+        );
+    }
+
+    private static void incrAndAddRt(MetricsKey metricsKey, MetricsPlaceType placeType, ServiceMetricsCollector<TimeCounterEvent> collector, TimeCounterEvent event) {
+        collector.increment(event.appName(), event.getAttachmentValue(ATTACHMENT_KEY_SERVICE), metricsKey,1);
```
Comment (songxiaosheng) [inline dubbo-metrics/dubbo-metrics-api/src/main/java/org/apache/dubbo/metrics/listener/MetricsServiceListener.java:52]: Can magic number 1 be modified to a constant

**Review by songxiaosheng — CHANGES_REQUESTED** [review]
There are only two small places

**Review by songxiaosheng — CHANGES_REQUESTED** [review]
There are only two small places

File: dubbo-metrics/dubbo-metrics-api/src/main/java/org/apache/dubbo/metrics/listener/MetricsServiceListener.java (hunk 1 above)
Comment (wxbty) [reply dubbo-metrics/dubbo-metrics-api/src/main/java/org/apache/dubbo/metrics/listener/MetricsServiceListener.java]: done

**Review by songxiaosheng — APPROVED** [review]
LGTM

**Review by songxiaosheng — APPROVED** [review]
LGTM

**Pull request merged**